node/pkg/processor/observation.go

```diff
@@ -160,55 +195,51 @@ func (p *Processor) handleObservation(ctx context.Context, obs *node_common.MsgW
 
 		observationsUnknownTotal.Inc()
 
-		p.state.signatures[hash] = &state{
+		s = &state{
 			firstObserved: time.Now(),
 			nextRetry:     time.Now().Add(nextRetryDuration(0)),
 			signatures:    map[common.Address][]byte{},
 			source:        "unknown",
 		}
-	}
 
-	p.state.signatures[hash].signatures[their_addr] = m.Signature
+		p.state.signatures[hash] = s
+	}
 
-	// Aggregate all valid signatures into a list of vaa.Signature and construct signed VAA.
-	agg := make([]bool, len(gs.Keys))
-	var sigs []*vaa.Signature
-	for i, a := range gs.Keys {
-		s, ok := p.state.signatures[hash].signatures[a]
+	s.signatures[their_addr] = m.Signature
 
-		if ok {
-			var bs [65]byte
-			if n := copy(bs[:], s); n != 65 {
-				panic(fmt.Sprintf("invalid sig len: %d", n))
-			}
-
-			sigs = append(sigs, &vaa.Signature{
-				Index:     uint8(i),
-				Signature: bs,
-			})
-		}
+	quorum := vaa.CalculateQuorum(len(gs.Keys))
 
-		agg[i] = ok
+	if len(s.signatures) < quorum {
+		// no quorum yet, we're done here
+		p.logger.Debug("quorum not yet met",
+			zap.String("digest", hash),
+			zap.String("messageId", m.MessageId),
+		)
+		return
 	}
+	// Now we *may* have quorum, depending on the guardian set in use.
+	// Let's construct the VAA and check if we actually have quorum.
 
-	if p.state.signatures[hash].ourObservation != nil {
+	sigsVaaFormat, agg := signaturesToVaaFormat(s.signatures, gs.Keys)
```

Comment:
I do not like this change. Please either revert it or move it to a separate PR that can be reviewed / discussed individually.

Reply:
I assume you're referring to https://github.com/wormhole-foundation/wormhole/pull/3197/commits/cedf65b4267ef3cf36f5aa88a77a131740f0b91d? 

This commit 
1. moves some code to a helper function signaturesToVaaFormat() for increased readability. No changes to the code. 
2. Adds an early return `if len(s.signatures) < quorum {return}`

Why do you not like it?

Reply:
A) I don't see the need to create a helper function that is only called once. Just leave the code inline like it was.
B) Since you are touching the code, we should drop `agg` all together since it is only used in debugs.
C) You are computing `quorum` twice.
D) I am not convinced that that early quorum check is the equivalent to what is down below.
E) You don't really need to move the quorum check up if you move the building of `sigs` into the check for `ouObservation` being null.

Reply:
A) I created the helper function from because of @kcsongor 's feedback to make the code more readable and I have to agree that it improves readability. Keep your functions short and all that. 
B) I've found `agg` very useful for debugging in the past, I'd like to keep it. 
C and D) I've removed the second re-calculation of `quorum`. The quorum checks are indeed not equivalent. The first one is faster though and we can return early in most cases. This is the performance improvement. Constructing the VAA actually takes a lot of time in the benchmarks and skipping that in most cases is where most of the performance improvement in this PR is coming from. I added additional comments to explain this. 
E) I've moved the call to `signaturesToVaaFormat` into the `if s.ourObservation != nil` branch. Good call!